File: flutter_frontend/lib/services/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;
import '../models/chat_message.dart';
import '../models/participant.dart';

class SocketService {
  static final SocketService _instance = SocketService._internal();
  factory SocketService() => _instance;
  SocketService._internal();

  IO.Socket? _socket;
  bool _isConnected = false;

  // Event callbacks
  Function(List<Participant>)? onParticipantsUpdate;
  Function(ChatMessage)? onChatMessage;
  Function(Participant)? onParticipantJoined;
  Function(String)? onParticipantLeft;
  Function()? onCourseEnded;

  bool get isConnected => _isConnected;

  void connect() {
    if (_socket != null && _isConnected) {
      return;
    }

    _socket = IO.io('http://localhost:3000', <String, dynamic>{
      'transports': ['websocket'],
      'autoConnect': false,
    });

    _socket!.connect();

    _socket!.on('connect', (_) {
      print('Socket.IO connected');
      _isConnected = true;
    });

    _socket!.on('disconnect', (_) {
      print('Socket.IO disconnected');
      _isConnected = false;
    });

    _socket!.on('error', (error) {
      print('Socket.IO error: $error');
    });

    // Meeting room events
    _socket!.on('participant-joined', (data) {
      if (onParticipantJoined != null) {
        onParticipantJoined!(Participant.fromJson(data));
      }
    });

    _socket!.on('participant-left', (data) {
      if (onParticipantLeft != null) {
        onParticipantLeft!(data['participantId']);
      }
    });

    _socket!.on('participants-update', (data) {
      if (onParticipantsUpdate != null) {
        final participants = (data as List)
            .map((p) => Participant.fromJson(p))
            .toList();
        onParticipantsUpdate!(participants);
      }
    });

    _socket!.on('chat-message', (data) {
      if (onChatMessage != null) {
        onChatMessage!(ChatMessage.fromJson(data));
      }
    });

    _socket!.on('course-ended', (_) {
      if (onCourseEnded != null) {
        onCourseEnded!();
      }
    });
  }

  void disconnect() {
    if (_socket != null) {
      _socket!.disconnect();
      _socket = null;
      _isConnected = false;
    }
  }

  void joinMeetingRoom(
    String meetingCode,
    String participantName,
    String participantId,
    bool isHost,
  ) {
    if (_socket != null && _isConnected) {
      _socket!.emit('join-meeting', {
        'meetingCode': meetingCode,
        'participantName': participantName,
        'participantId': participantId,
        'isHost': isHost,
      });
    }
  }

  void leaveMeetingRoom(String meetingCode, String participantId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('leave-meeting', {
        'meetingCode': meetingCode,
        'participantId': participantId,
      });
    }
  }

  void sendChatMessage(String meetingCode, String sender, String message) {
    if (_socket != null && _isConnected) {
      _socket!.emit('chat-message', {
        'meetingCode': meetingCode,
        'sender': sender,
        'message': message,
      });
    }
  }

  void toggleAudio(String meetingCode, String participantId, bool enabled) {
    if (_socket != null && _isConnected) {
      _socket!.emit('toggle-audio', {
        'meetingCode': meetingCode,
        'participantId': participantId,
        'enabled': enabled,
      });
    }
  }

  void toggleVideo(String meetingCode, String participantId, bool enabled) {
    if (_socket != null && _isConnected) {
      _socket!.emit('toggle-video', {
        'meetingCode': meetingCode,
        'participantId': participantId,
        'enabled': enabled,
      });
    }
  }
}

File: frontend/src/services/socketService.ts
import { io, Socket } from 'socket.io-client';

export interface ChatMessage {
  id: string;
  sender: string;
  message: string;
  timestamp: string;
  meetingCode?: string;
}

export interface ParticipantEvent {
  participantId: string;
  participantName: string;
  isHost?: boolean;
  joinedAt?: string;
  leftAt?: string;
  meetingCode?: string;
}

class SocketService {
  private socket: Socket | null = null;
  private isConnected = false;

  connect(): void {
    if (this.socket?.connected) {
      return;
    }

    // Use environment variable or default to backend port
    const backendUrl = process.env.REACT_APP_BACKEND_URL || 'http://localhost:3000';
    this.socket = io(backendUrl, {
      transports: ['websocket', 'polling'],
      timeout: 10000,
    });

    this.socket.on('connect', () => {
      console.log('✅ Connected to Socket.IO server');
      this.isConnected = true;
    });

    this.socket.on('disconnect', () => {
      console.log('❌ Disconnected from Socket.IO server');
      this.isConnected = false;
    });

    this.socket.on('connect_error', (error) => {
      console.error('❌ Socket.IO connection error:', error);
      this.isConnected = false;
    });
  }

  disconnect(): void {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
      this.isConnected = false;
    }
  }

  // Join a meeting room for real-time updates
  joinMeetingRoom(meetingCode: string, participantName: string, participantId?: string, isHost?: boolean): void {
    if (this.socket) {
      this.socket.emit('join-meeting', {
        meetingCode,
        participantId: participantId || Date.now().toString(), // Generate ID if not provided
        participantName,
        isHost: isHost || false,
        timestamp: new Date().toISOString(),
      });
    }
  }

  // Leave a meeting room
  leaveMeetingRoom(meetingCode: string, participantName: string): void {
    if (this.socket) {
      this.socket.emit('leave-meeting', {
        meetingCode,
        participantName,
        timestamp: new Date().toISOString(),
      });
    }
  }

  // Send a chat message
  sendChatMessage(meetingCode: string, sender: string, message: string): void {
    if (this.socket) {
      const chatMessage: ChatMessage = {
        id: Date.now().toString(),
        sender,
        message,
        timestamp: new Date().toISOString(),
        meetingCode,
      };
      
      this.socket.emit('chat-message', chatMessage);
    }
  }

  // Listen for chat messages
  onChatMessage(callback: (message: ChatMessage) => void): void {
    if (this.socket) {
      this.socket.on('chat-message', callback);
    }
  }

  // Listen for participant joined events
  onParticipantJoined(callback: (data: ParticipantEvent) => void): void {
    if (this.socket) {
      this.socket.on('participant-joined', callback);
    }
  }

  // Listen for participant left events
  onParticipantLeft(callback: (data: ParticipantEvent) => void): void {
    if (this.socket) {
      this.socket.on('participant-left', callback);
    }
  }

  // Listen for current participants list
  onCurrentParticipants(callback: (participants: any[]) => void): void {
    if (this.socket) {
      this.socket.on('current-participants', callback);
    }
  }

  // Listen for error events
  onError(callback: (error: any) => void): void {
    if (this.socket) {
      this.socket.on('error', callback);
    }
  }

  // Remove all listeners for a specific event
  removeListener(event: string): void {
    if (this.socket) {
      this.socket.off(event);
    }
  }

  // Remove all listeners
  removeAllListeners(): void {
    if (this.socket) {
      this.socket.removeAllListeners();
    }
  }

  getConnectionStatus(): boolean {
    return this.isConnected;
  }

  getSocket(): Socket | null {
    return this.socket;
  }

  // WebRTC signaling methods
  sendOffer(offer: RTCSessionDescriptionInit, to?: string): void {
    if (this.socket) {
      this.socket.emit('offer', { offer, to });
    }
  }

  sendAnswer(answer: RTCSessionDescriptionInit, to?: string): void {
    if (this.socket) {
      this.socket.emit('answer', { answer, to });
    }
  }

  sendIceCandidate(candidate: RTCIceCandidate, to?: string): void {
    if (this.socket) {
      this.socket.emit('ice-candidate', { candidate, to });
    }
  }

  // Media control events
  toggleAudio(enabled: boolean): void {
    if (this.socket) {
      this.socket.emit('toggle-audio', { enabled });
    }
  }

  toggleVideo(enabled: boolean): void {
    if (this.socket) {
      this.socket.emit('toggle-video', { enabled });
    }
  }

  // Screen sharing events
  startScreenShare(): void {
    if (this.socket) {
      this.socket.emit('start-screen-share', { timestamp: new Date().toISOString() });
    }
  }

  stopScreenShare(): void {
    if (this.socket) {
      this.socket.emit('stop-screen-share', { timestamp: new Date().toISOString() });
    }
  }

  sendScreenShareOffer(offer: RTCSessionDescriptionInit): void {
    if (this.socket) {
      this.socket.emit('screen-share-offer', { offer });
    }
  }

  sendScreenShareAnswer(answer: RTCSessionDescriptionInit): void {
    if (this.socket) {
      this.socket.emit('screen-share-answer', { answer });
    }
  }

  sendScreenShareIceCandidate(candidate: RTCIceCandidate): void {
    if (this.socket) {
      this.socket.emit('screen-share-ice-candidate', { candidate });
    }
  }

  // Event listeners for WebRTC signaling
  onOffer(callback: (data: any) => void): void {
    if (this.socket) {
      this.socket.on('offer', callback);
    }
  }

  onAnswer(callback: (data: any) => void): void {
    if (this.socket) {
      this.socket.on('answer', callback);
    }
  }

  onIceCandidate(callback: (data: any) => void): void {
    if (this.socket) {
      this.socket.on('ice-candidate', callback);
    }
  }

  onParticipantAudioToggle(callback: (data: any) => void): void {
    if (this.socket) {
      this.socket.on('participant-audio-toggle', callback);
    }
  }

  onParticipantVideoToggle(callback: (data: any) => void): void {
    if (this.socket) {
      this.socket.on('participant-video-toggle', callback);
    }
  }

  onScreenShareStarted(callback: (data: any) => void): void {
    if (this.socket) {
      this.socket.on('screen-share-started', callback);
    }
  }

  onScreenShareStopped(callback: (data: any) => void): void {
    if (this.socket) {
      this.socket.on('screen-share-stopped', callback);
    }
  }

  onScreenShareOffer(callback: (data: any) => void): void {
    if (this.socket) {
      this.socket.on('screen-share-offer', callback);
    }
  }

  onScreenShareAnswer(callback: (data: any) => void): void {
    if (this.socket) {
      this.socket.on('screen-share-answer', callback);
    }
  }

  onScreenShareIceCandidate(callback: (data: any) => void): void {
    if (this.socket) {
      this.socket.on('screen-share-ice-candidate', callback);
    }
  }
}

// Export a singleton instance
export const socketService = new SocketService();
export default socketService;